apps/server/lib/lexical/server/project/progress/state.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Lexical.Server.Project.Progress.State do
+  alias Lexical.Project
+  alias Lexical.Protocol.Notifications
+  alias Lexical.Server.Project.Progress.Value
+  alias Lexical.Server.Transport
+
+  import Lexical.RemoteControl.Api.Messages
+
+  defstruct project: nil, progress_by_label: %{}
+
+  def new(%Project{} = project) do
+    %__MODULE__{project: project}
+  end
+
+  def begin(%__MODULE__{} = state, project_progress(label: label)) do
+    token = System.unique_integer([:positive])
+    progress = Value.new(token, :begin, label)
+    progress_by_label = Map.put(state.progress_by_label, label, progress)
+
+    write_work_done(token)
+    write(progress)
+
+    %{state | progress_by_label: progress_by_label}
+  end
+
+  def report(%__MODULE__{} = state, project_progress(label: label, message: message)) do
+    token = get_token(state, label)
+
+    progress = Value.new(token, :report, message)
+    write(progress)
+
+    progress_by_label = Map.put(state.progress_by_label, label, progress)
+    %{state | progress_by_label: progress_by_label}
+  end
+
+  def complete(%__MODULE__{} = state, project_progress(label: label, message: message)) do
+    token = get_token(state, label)
+
+    progress = Value.new(token, :end, message)
+    write(progress)
+
+    clear_progress_by_label(state, label)
```

Comment:
same comment here:
```suggestion
    {progress, progress_by_label} = 
      Map.get_and_update(state.progress_by_label, label, fn _ -> :pop end)
    progress
    |> Value.finish()
    |> write()
   %{state | progress_by_label: progress_by_label}
```

This will remove the need for the `clear_progress_by_label`